### ISSUE TITLE
leaflet.css - Unexpected duplicate selector (SonarQube Major Code Smell x8)

### ISSUE THREAD

**Issue by jspraul** [Leaflet/Leaflet#8121]
- [x] I'm sure this is a Leaflet code issue, not an issue with my own code nor with the framework I'm using (Cordova, Ionic, Angular, React…)
- [x] I've searched through the issues to make sure it's not yet reported

**Steps to reproduce**
- run [SonarQube](https://www.sonarqube.org/) static code analysis on leaflet.css

**Expected behavior**
No (simple) SonarQube code smells causing default Quality Gate to fail

**Current behavior**
8 violations of SonarQube [rule 4666](https://rules.sonarsource.com/css/RSPEC-4666) which has Major severity:
Unexpected duplicate selector ".leaflet-container", first used at line 17 L72
Unexpected duplicate selector ".leaflet-control", first used at line 122 L146
Unexpected duplicate selector ".leaflet-right .leaflet-control", first used at line 150 L162
Unexpected duplicate selector ".leaflet-zoom-anim .leaflet-zoom-animated", first used at line 186 L189
Unexpected duplicate selector ".leaflet-container", first used at line 17 L252
Unexpected duplicate selector ".leaflet-container a", first used at line 75 L256
Unexpected duplicate selector ".leaflet-zoom-box", first used at line 85 L262
Unexpected duplicate selector ".leaflet-container", first used at line 17 L269

**Environment**
- Leaflet version: 1.7.1
- Browser (with version): N/A (Google Chrome Version 100.0.4896.60 (Official Build) (64-bit))
- OS/Platform (with version): Windows 10 Pro

**Additional context**
I really appreciate what this project offers and wanted to take the time to document this in case combining the selectors is possible upstream. The leaflet.css file could be minified but perhaps this could be addressed in the default download.

**Minimal example reproducing the issue**
I've included the errors reported by SonarQube which include for example L72, indicating a selector is repeated on line 72 in leaflet.css. I believe the fix would be to merge the CSS properties with the initial selector, in this case on line 17.

**Comment by Malvoz** [comment]
In some cases selectors are duplicated for developer convenience, for readability, for example:

https://github.com/Leaflet/Leaflet/blob/7847cc6913c296fb7e409e086a57366237c42728/dist/leaflet.css#L253-L255

https://github.com/Leaflet/Leaflet/blob/7847cc6913c296fb7e409e086a57366237c42728/dist/leaflet.css#L268-L269

In some cases not:

https://github.com/Leaflet/Leaflet/blob/7847cc6913c296fb7e409e086a57366237c42728/dist/leaflet.css#L17-L18

https://github.com/Leaflet/Leaflet/blob/7847cc6913c296fb7e409e086a57366237c42728/dist/leaflet.css#L77-L78

Is it worth deduplicating only some of these, and keep other duplicates to preserve useful comments?

Instead, perhaps comments could be stripped, and duplicate selectors merged, in the process of minification: https://github.com/Leaflet/Leaflet/issues/4737.

**Comment by pto1026** [comment]
Can you assign me to this issue, I'd like to tackle the refactoring

**Comment by Malvoz** [comment]
@pto1026 I believe the refactoring would potentially happen in 2 parts, as described above, see the issues:

> - Duplicate selectors could be avoided through linting: https://github.com/Leaflet/Leaflet/issues/7806
> - Comments could be stripped in the process of minification: https://github.com/Leaflet/Leaflet/issues/4737

Or were you thinking of removing comments and deduplicating selectors manually?

**Comment by TomerPacific** [comment]
@Falke-Design  - Seeing as this issue hasn't been handled, I'd like to help out.
What would be the best approach to identify which duplicated selectors we would like to keep and which we would like to get rid of?

**Comment by jonkoops** [comment]
I don't really see the issue in having duplicate selectors for code organization. Should we perhaps just close this issue?

**Comment by jspraul** [comment]
Many developers are required to pass SonarQube's default configuration, through not fault of their own.

It's not a big deal in any way from a code organization perspective or really any perspective, just an opportunity to make their lives easier.

The scenario is: ➀ project is sailing through CI, ➁ add Leaflet, ➂ not smooth sailing anymore.

**Comment by jonkoops** [comment]
I really don't understand why folks have made SonarCube their supreme deity of code quality, but sure. I'm not principally against this change, but I'd like other maintainers lean in.

**Comment by mourner** [comment]
Yeah, I feel like this is SonarCube problem, no idea why some arbitrary rules that don't even make much sense (why are external libraries linted? Why are duplicate selectors a "Major severity"?) have to force maintainers to adjust.

As @Malvoz mentioned, there are some duplicate selector cases we could fix, but some should be left for code organization / readability.
